File: app/src/main/java/com/android_cademy/starwarstinder/model/ProfileRepository.java
package com.android_cademy.starwarstinder.model;

import androidx.lifecycle.LiveData;

import com.android_cademy.starwarstinder.db.AppDatabase;
import com.android_cademy.starwarstinder.network.ProfileNetworkDataSource;

import java.util.List;

import javax.inject.Inject;

public class ProfileRepository {

    private static int MIN_PROFILES_TO_CACHE = 10;
    private final AppDatabase db;
    private final ProfileNetworkDataSource profileNetworkDataSource;

    @Inject
    public ProfileRepository(final AppDatabase db, ProfileNetworkDataSource dataSource) {
        this.db = db;
        this.profileNetworkDataSource = dataSource;

        LiveData<List<Profile>> networkData = dataSource.getCurrentProfiles();
        networkData.observeForever(profilesFromNetwork -> new Thread(() -> {
            deleteOldData();
            Profile[] profile = profilesFromNetwork.toArray(new Profile[profilesFromNetwork.size
                    ()]);
            db.profileDao().bulkInsert(profile);
        }).start());
    }

    private void deleteOldData() {
        db.profileDao().deleteProfiles();
    }

    public LiveData<List<Profile>> getProfiles() {
        initializeData();
        return db.profileDao().getProfiles();
    }

    private void initializeData() {
        //regular thread is used to simplify the example. don't do that in the production :)
        new Thread(() -> {
            if (isFetchNeeded()) {
                startFetchProfilesService();
            }
        }).start();
    }

    private void startFetchProfilesService() {
        profileNetworkDataSource.startFetchProfiles();
    }

    private boolean isFetchNeeded() {
        int count = db.profileDao().countProfiles();
        return count <= ProfileRepository.MIN_PROFILES_TO_CACHE;
    }
}

File: app/src/main/java/com/android_cademy/starwarstinder/model/ProfileRepository.kt
package com.android_cademy.starwarstinder.model

import androidx.lifecycle.LiveData

import com.android_academy.network.ProfileNetworkDataSource

import javax.inject.Inject

class ProfileRepository @Inject
constructor(private val db: com.android_academy.db.AppDatabase, private val profileNetworkDataSource: ProfileNetworkDataSource) {

    val profiles: LiveData<List<com.android_academy.db.Profile>>
        get() {
            return db.profileDao().getAll()
        }

    private val isFetchNeeded: Boolean
        get() {
            val count = db.profileDao().countProfiles()
            return count <= MIN_PROFILES_TO_CACHE
        }

    init {
        val networkData = profileNetworkDataSource.networkProfiles
        networkData.observeForever { profilesFromNetwork ->
            Thread {
                deleteOldData()
                val profile = profilesFromNetwork.toTypedArray()
                db.profileDao().bulkInsert(*profile)
            }.start()
        }
        fetchData()
    }

    private fun deleteOldData() {
        db.profileDao().deleteProfiles()
    }

    private fun fetchData() {
        //regular thread is used to simplify the example. don't do that in the production :)
        Thread {
            if (isFetchNeeded) {
                startFetchProfilesService()
            }
        }.start()
    }

    private fun startFetchProfilesService() {
        profileNetworkDataSource.startFetchProfiles()
    }

    companion object {
        private val MIN_PROFILES_TO_CACHE = 10
    }
}
